File Manager and Gantt - Add splitter and extract common parts of its functionality

Staziz opened the pull request:
<!--
Make sure that you've read the "Contributing Code and Content" section in CONTRIBUTING.md

If you are submitting a bug fix, the subject should contain "fixes ISSUE_ID" or "resolves ISSUE_ID".

In addition, please clarify:

1. What did you do? 
2. How did you do it?
3. How should we verify this?

-->

Commit:
File Manager and Gantt - Add splitter and extract common parts of its functionality

Commit:
Splitter - Add compatibility theme file

maxerokh commented on styles/widgets/common/drawer.less:
```diff
@@ -11,6 +11,7 @@
 }
 
 .dx-drawer-panel-content {
+    display: inline-flex;
```
Move to file manager and gantt CSS classes selectors

maxerokh commented on js/ui/splitter.js:
```diff
@@ -0,0 +1,96 @@
+        this._container = this.option("container");
+        this._leftElement = this.option("leftElement");
+        this._rightElement = this.option("rightElement");
+        this._onSplitterChanged = this._createActionByOption("onSplitterChanged");
```
this._onSplitterChanged -> this._onApplyPanelSize

maxerokh commented on js/ui/splitter.js:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    computeRightPanelWidth(leftPanelWidth) {
+        return this._maxAvailablePosX - leftPanelWidth - this._$splitterBorder.width();
```
Use the this._container.width() instead of this._maxAvailablePosX

maxerokh commented on js/ui/splitter.js:
```diff
@@ -0,0 +1,96 @@
+        if(!this._isSplitterActive) {
+            return;
+        }
+        this._onSplitterChanged(this._calculateLeftElementWidth(e));
```
compute... calculate... Choose one naming format.

maxerokh commented on js/ui/file_manager/ui.file_manager.adaptivity.js:
```diff
@@ -27,15 +28,43 @@ class FileManagerAdaptivityControl extends Widget {
     }
 
+    _onSplitterChanged(elementsWidth) {
+        this._updateWidth(elementsWidth.actionValue);
```
Do you use the `_updateWidth` method in another place? Do you have reasons to create the method?

Commit:
Splitter - Tests fix

Staziz commented:
Changed things that was offered, and hope that tests are up now

Staziz commented:
Merged some recent changes in themebuilder

Commit:
Function naming is now in in common style

Staziz commented:
Function naming is now in common style

Commit:
Removed some unused variables

Staziz commented:
Removed some unused variables

Staziz commented:
Merged with latest project updates

Commit:
File Manager - Adaptivity fix

Gantt - Columns resizing fix (still have bug of horizontal scrollbar in rightmost position of splitter)
Splitter - By default separates elements in percent, but start value is still set in pixels

Staziz commented:
File Manager - Adaptivity fix
Gantt - Columns resizing fix (still have bug of horizontal scrollbar in rightmost position of splitter)
Splitter - By default separates elements in percent, but start value is still set in pixels

Commit:
Reworked splitter

Staziz commented:
Just another try to make this splitter great (again?)

Commit:
WidgetsCreation tests fix

Commit:
Add missed fileManager test

vladkovl commented on js/ui/file_manager/ui.file_manager.adaptivity.js:
```diff
@@ -27,15 +28,42 @@ class FileManagerAdaptivityControl extends Widget {
+    _createDrawerTemplate(container) {
+        this.option("drawerTemplate")(container);
+        const leftElement = container;
+        const rightElement = this.$element().find(".dx-drawer-content");
```
I think this is a bad idea to modify the internal rendering of a Drawer widget. Can we use the Drawer's public API or modify only our own render?

vladkovl commented on styles/widgets/common/fileManager.less:
```diff
@@ -135,12 +135,19 @@
             padding: 5px 10px;
+            height: 100%;
+            min-width: 250px;
+            max-width: unset;
```
How good it looks on wide screens?

Staziz replied:
By default dirs panel is rendered with 250px width, as far as I could explore on my only monitor.

vladkovl commented on styles/widgets/generic/splitter.generic.less:
```diff
@@ -0,0 +1,12 @@
+.dx-splitter {
+    width: 4px;
+    background-color: @datagrid-column-separator-bg;
```
Do not use constants from other widgets, because their values can be changed by the customer in the process of creating new themes using a theme builder tool.

Commit:
Corrected found errors

Commit:
Fix for some little mistakes

Commit:
Relative widths are now more precise

Staziz commented:
Relative widths are now more precise

Commit:
File Manager tests fix

Commit:
Code is refactored a bit

vladkovl commented on js/ui/splitter.js:
```diff
@@ -0,0 +1,163 @@
+        eventsEngine.on(this._$splitter, SPLITTER_POINTER_DOWN_EVENT_NAME, this._onMouseDownHandler.bind(this));
+        eventsEngine.on(document, SPLITTER_POINTER_MOVE_EVENT_NAME, this._onMouseMoveHandler.bind(this));
+        eventsEngine.on(document, SPLITTER_POINTER_UP_EVENT_NAME, this._onMouseUpHandler.bind(this));
+        eventsEngine.on(getWindow(), SPLITTER_WINDOW_RESIZE_EVENT_NAME, this._windowResizeHandler.bind(this));
```
Widget have already a subscription to the window.resize event. And it has a `_dimensionChanged` as a handler for this event. I think we should use this method instead of the new one.

vladkovl commented on styles/widgets/common/fileManager.less:
```diff
@@ -135,15 +135,27 @@
+            min-width: 250px;
+            max-width: unset;
+            width: 100%;
+            display: inline-flex;
```
Why can't we just use a `flex` instead of the `inline-flex`?

vladkovl commented on styles/widgets/generic/color-schemes/contrast/generic.contrast.less:
```diff
@@ -1641,3 +1641,7 @@
+
+// Splitter
+@splitter-border-color: @base-border-color;
+@splitter-bg: darken(@datagrid-base-background-color, 13.5%);
```
Don't use datagrid color

vladkovl commented on js/ui/splitter.js:
```diff
@@ -0,0 +1,163 @@
+        return this._getElementMinMaxWidthRecursiveCore(this._$leftElement[0], "maxWidth");
+    }
+
+    _getElementMinMaxWidthRecursiveCore(element, minMaxAttr) {
```
Do we really need this checks? Can we use a solution based only on CSS?

Commit:
Improved splitter behavior

Add splitter test for gantt

Staziz commented:
Improved splitter behavior, added splitter test for gantt, DON'T added filemanager test, it's causing some problems

Commit:
Adds File Manager tests and performs some requested changes

vladkovl commented on js/ui/file_manager/ui.file_manager.adaptivity.js:
```diff
@@ -10,6 +10,8 @@ import SplitterControl from "../splitter";
 const window = getWindow();
 const ADAPTIVE_STATE_SCREEN_WIDTH = 573;
 
+const DRAWER_PANEL_CONTENT_INITIAL = "dx-drawer-panel-content-initial";
```
I think that the class name `"dx-drawer-panel-content-initial"` should have the prefix `"dx-filemanager"`, because this class name is used for the FileManager widget, and not for a standalone Drawer widget.

vladkovl commented on testing/tests/DevExpress.ui.widgets/fileManagerParts/navigation.tests.js:
```diff
@@ -151,4 +152,41 @@ QUnit.module("Navigation operations", moduleConfig, () => {
     });
 
+    test("splitter should change width of dirs tree and file items areas", function(assert) {
+        renderer.fn.width = function() {
```
You need to restore the original function `width()` at the end of the test, otherwise other tests will use this modified function and they will fail because of this.

Commit:
Fix for File Manager tests
